### PR TITLE
Argon2 Android library.

### PR DESCRIPTION
**Pull request by alan-signal**
This is the Android Argon2 library.

It gives a wrapper around the native argon2 calls and has testcases ported from the c library.

I believe this repo could serve as a place for both Android and iOS to place their library code.

Example usage:
```
Argon2 argon2 = new Argon2.Builder()
                          .version(ARGON2_VERSION_13)
                          .type(Argon2id)
                          .iterations(2)
                          .memory(16)
                          .parallelism(2)
                          .hashLength(32)
                          .build();

Argon2.Result result = argon2.hash(passwordBytes, saltBytes);
byte[]        hash   = result.getHash();
```

### CODE REVIEW

File: android/argon2/jni/org_signal_argon2_Argon2Native.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+   jbyteArray jHash,
+   jobject jEncoded,
+   jint argon_type,
+   jint version)
+{
+  jsize  pwd_size     = (*env)->GetArrayLength(env, jPwd);
+  jsize  salt_size    = (*env)->GetArrayLength(env, jSalt);
+  jsize  outLen       = (*env)->GetArrayLength(env, jHash);
+  jbyte* pwdElements  = (*env)->GetByteArrayElements(env, jPwd, NULL);
+  jbyte* saltElements = (*env)->GetByteArrayElements(env, jSalt, NULL);
```
Comment (geogriff-signal) [inline android/argon2/jni/org_signal_argon2_Argon2Native.c]: Also need exception checking all around, as the [JNI spec](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html) states: "After an exception has been raised, the native code must first clear the exception before making other JNI calls. When there is a pending exception, the JNI functions that are safe to call are..."

File: android/argon2/jni/org_signal_argon2_Argon2Native.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+JNIEXPORT jint JNICALL Java_org_signal_argon2_Argon2Native_verify
+  (JNIEnv *env, jclass clazz, jstring jEncoded, jbyteArray jPwd, jint argon_type)
+{
+  const char  *encoded  = (*env)->GetStringUTFChars(env, jEncoded, NULL);
+        jsize  pwd_size = (*env)->GetArrayLength(env, jPwd);
+        jbyte *pwd      = (*env)->GetByteArrayElements(env, jPwd, NULL);
+
+  int ret = argon2_verify((char *)encoded, pwd, pwd_size, argon_type);
```
Comment (geogriff-signal) [inline android/argon2/jni/org_signal_argon2_Argon2Native.c]: argon2_verify is expecting a NULL-terminated string for `encoded`, and `GetStringUTFChars` apparently isn't guaranteed to return such.

File: android/argon2/jni/org_signal_argon2_Argon2Native.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   jint argon_type,
+   jint version)
+{
+  jsize  pwd_size     = (*env)->GetArrayLength(env, jPwd);
+  jsize  salt_size    = (*env)->GetArrayLength(env, jSalt);
+  jsize  outLen       = (*env)->GetArrayLength(env, jHash);
+  jbyte* pwdElements  = (*env)->GetByteArrayElements(env, jPwd, NULL);
+  jbyte* saltElements = (*env)->GetByteArrayElements(env, jSalt, NULL);
+
+  unsigned char out[outLen];
```
Comment (geogriff-signal) [inline android/argon2/jni/org_signal_argon2_Argon2Native.c]: please add at least an assert here to sanity check outLen so you don't accidentally allocate a huge stack variable and overflow.

File: android/argon2/jni/org_signal_argon2_Argon2Native.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  int ret = argon2_hash(t, m, parallelism,
+                        pwdElements, pwd_size,
+                        saltElements, salt_size,
+                        out, outLen,
+                        encoded, ENCODED_LEN,
+                        argon_type,
+                        version);
+
+  (*env)->ReleaseByteArrayElements(env, jPwd, pwdElements, JNI_ABORT);
```
[the remark this answers is not in the document]
Comment (alan-signal) [reply android/argon2/jni/org_signal_argon2_Argon2Native.c]: Performance, it says I don't want to copy back any changes to these arrays back to Java heap.  https://developer.android.com/training/articles/perf-jni#region-calls

File: android/argon2/jni/org_signal_argon2_Argon2Native.c (hunk 2 above)
Comment (alan-signal) [reply android/argon2/jni/org_signal_argon2_Argon2Native.c]: I know the spec doesn't say it explicitly, but it is how it works and every example you can find will show an assumption of null termination.

File: android/argon2/jni/org_signal_argon2_Argon2Native.c (hunk 3 above)
Comment (alan-signal) [reply android/argon2/jni/org_signal_argon2_Argon2Native.c]: `outLen` was not passed in to the call, it's the length of a supplied array:

```
jsize  outLen       = (*env)->GetArrayLength(env, jHash);
```

But now I'm wondering why I didn't just use `getByteArrayElements` and copy back to Java.

**Pull request merged**